### PR TITLE
Migrate corpus pruning [3/N]

### PR DESCRIPTION
**Pull request by alhijazi**
Move `process_corpus_crashes` to postprocess:
1- From `utask_main`: Pack the crash.unit_path into a zip file and upload using a signed url
2- From `utask_postprocess`: get the zip, unpack individual files and upload necessary ones to blob storage, store the blob key in the created Testcase's `fuzzed_keys` attribute.

### CODE REVIEW

File: src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py [hunk 1]
```diff
@@ -729,42 +731,76 @@ def do_corpus_pruning(context, revision):
...
+  file_host.copy_file_from_worker(crash.unit_path, unit_path)
+  crash.unit_path = unit_path
+
+
+def _upload_corpus_crashes_zip(context, result, corpus_crashes_upload_url):
+  """Packs the corpus crashes in a zip file. The file is then uploaded
+  using the signed upload url from the input."""
+  with tempfile.NamedTemporaryFile(delete=True) as f:
+    zip_filename = f.name
+    zip_file = zipfile.ZipFile(zip_filename, 'w')
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py:755]: Can we use `with` to open this file?

File: src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py [hunk 2]
```diff
@@ -729,42 +731,76 @@ def do_corpus_pruning(context, revision):
...
+      fuzz_target.project_qualified_name(), crash_revision)
+
+  # Copy the crashes zip file from cloud storage into a temporary directory.
+  temp_dir = tempfile.mkdtemp()
+  corpus_crashes_zip_local_path = os.path.join(temp_dir, 'corpus_crashes.zip')
+  storage.copy_file_from(
+      blobs.get_gcs_path(output.uworker_input.corpus_pruning_task_input.
+                         corpus_crashes_blob_name),
+      corpus_crashes_zip_local_path)
+  crashes_zip_file = zipfile.ZipFile(corpus_crashes_zip_local_path)
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py:787]: Can we use archive module and/or `with` syntax?

File: src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py [hunk 3]
```diff
@@ -729,42 +731,76 @@ def do_corpus_pruning(context, revision):
...
         project_name,
         crash.crash_type,
         crash.crash_state,
         crash.security_flag,
-        fuzz_target=context.fuzz_target.project_qualified_name())
+        fuzz_target=fuzz_target.project_qualified_name())
     if existing_testcase:
       continue
 
+    crash_local_unit_path = os.path.join(temp_dir, crash.unit_path.lstrip('/'))
```
Comment (jonathanmetzman) [inline src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py:799]: Why do we need this hacky looking l-strip?

File: src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py [hunk 4]
```diff
@@ -729,42 +731,76 @@ def do_corpus_pruning(context, revision):
...
+
+def _upload_corpus_crashes_zip(context, result, corpus_crashes_upload_url):
+  """Packs the corpus crashes in a zip file. The file is then uploaded
+  using the signed upload url from the input."""
+  with tempfile.NamedTemporaryFile(delete=True) as f:
+    zip_filename = f.name
+    zip_file = zipfile.ZipFile(zip_filename, 'w')
+    for crash in result.crashes:
+      _update_crash_unit_path(context, crash)
+      zip_file.write(crash.unit_path, crash.unit_path, zipfile.ZIP_DEFLATED)
```
[the remark this answers is not in the document]
Comment (alhijazi) [reply src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py:758]: Not necessarily mandatory, but I was following the example of what's already used in clusterfuzz code. 
https://github.com/search?q=repo%3Agoogle%2Fclusterfuzz%20zipfile.ZIP_DEFLATED&type=code

File: src/clusterfuzz/_internal/protos/uworker_msg.proto [hunk 5]
```diff
@@ -166,6 +168,9 @@ message CrashInfo {
   optional string crash_type = 3;
   optional string crash_state = 4;
   optional bool security_flag = 5;
+  optional string unit_path = 6;
```
[the remark this answers is not in the document]
Comment (alhijazi) [reply src/clusterfuzz/_internal/protos/uworker_msg.proto]: I modified it to use only names, but I am worried about  having the same file name for different unit paths.
Could this happen?
unit_path1 = a/b/file_name
unit_path2 = c/d/file_name

**Review by jonathanmetzman — COMMENTED** [review]
LGTM

Do you want to merge this and deploy?
Then you should try to run the schedule_corpus_pruning k8s workload so this starts now.

**Review by jonathanmetzman — APPROVED** [review]

**Pull request merged**